Implement SCP Support

Qalthos opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

Fixes https://github.com/ansible/pylibssh/issues/151
Adds scp support to pylibssh.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

Qalthos committed:
Add infra to support scp

Qalthos committed:
Add scp files

Co-Authored-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

Qalthos committed:
Fix up issues

Qalthos committed:
Add changelog

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+
+    def __dealloc__(self):
+        if self._libssh_session is not NULL:
+            self._libssh_session = NULL
```
Assigning NULL if memory is allocated doesn't seem right, might result in memory leaks. Is the `__dealloc__` required here? The deallocation of `self._libssh_session` should be taken care of by the session object.

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+        remote_file_b = remote_file
+        if isinstance(remote_file_b, unicode):
+            remote_file_b = remote_file.encode("utf-8")
+            remote_dir_b, filename_b = os.path.split(remote_file_b)
```
Is this line required `remote_dir_b` and `filename_b` not used in function?

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+                raise LibsshSCPException("Memory allocation error")
+
+            # Indicate the transfer is ready to begin
+            libssh.ssh_scp_accept_request(scp)
```
Is return code checking required here as the function can return either `SSH_OK` or `SSH_ERROR `?

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+            # Read the file
+            rc = libssh.ssh_scp_read(scp, read_buffer, size)
+            if rc == libssh.SSH_ERROR:
+                PyMem_Free(read_buffer)
```
This line can be moved to `finally` block.

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+
+            with open(local_file, "w+") as f:
+                f.write(read_buffer.decode("utf-8"))
+            PyMem_Free(read_buffer)
```
Same as above, check is read_buffer is allocated before freeing in `finally` block

ganeshrn commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,132 @@
+        return libssh.SSH_OK
+
+    def close(self):
+        if self._libssh_session is not NULL:
```
I think this `close` method should handle freeing of `scp` resources that are still allocated.

Qalthos replied:
Hm, if `_libssh_session` is handled by `session`'s `__dealloc__`, I don't think there is anything left to free as the scp sessions seem to be single-purpose and get freed at the end of a call.

Is there anything I'm missing that needs to be here, or just drop it like `__dealloc__`?

ganeshrn replied:
Yeah if it not required we can drop it

ganeshrn reviewed (commented):
Please add unit tests to cover SCP scenarios.

ganeshrn reviewed (commented):
Also please update the user_guide https://github.com/ansible/pylibssh/blob/devel/docs/user_guide.rst

Qalthos committed:
Remove dealloc, close, and unused strings

Qalthos committed:
Free read_buffer in finally

Qalthos committed:
Check for error on scp_accept_request

Qalthos committed:
Properly write non-text files and set mode

Qalthos committed:
Add scp unit tests

webknjaz commented:
I'll rebase this to pull in the fixes from devel.

webknjaz reviewed (changes requested):
I've added the linter fixes and rebased on top of devel and now it's clear that the test you added (`tests/unit/scp_test.py::test_put`) fails: https://github.com/ansible/pylibssh/pull/157/checks?check_run_id=1504712459#step:13:235. Blocking until solved.

Qalthos committed:
Limit file permission to lower nine bits

Qalthos committed:
Remote file no longer needs a directory

Qalthos committed:
Update user guide

Qalthos commented:
I updated `user_docs` and split the included files into a few domain-specific versions to keep them a bit easier to follow if someone wants to look at the files directly.

ganeshrn reviewed (approved):
Thanks for working on this.

webknjaz commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,141 @@
+        self.session = session
+        self._libssh_session = get_libssh_session(session)
+
+    def put(self, local_file, remote_file):
```
Could you add a docstring?

webknjaz commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,141 @@
+
+            return libssh.SSH_OK
+
+    def get(self, remote_file, local_file):
```
This needs a docstring too.

Qalthos committed:
Add docstrings to put and get

webknjaz commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,159 @@
+
+    def put(self, local_file, remote_file):
+        """
+        Create an SCP channel and send a file to the remote host over that channel
```
PEP 257 fix:
```suggestion
        Create an SCP channel and send a file to the remote host over that channel.
```

webknjaz commented on src/pylibsshext/scp.pyx:
```diff
@@ -0,0 +1,159 @@
+
+    def get(self, remote_file, local_file):
+        """
+        Create an SCP channel and retrieve a file from the remote host over that channel
```
PEP257:
```suggestion
        Create an SCP channel and retrieve a file from the remote host over that channel.
```
